### PR TITLE
Geom: Integrate new `Line` class

### PR DESCRIPTION
**Pull request by jsn0716**
### Description

Integrates the new `Line` class, replacing the old one.

A couple things to note:
1. As per the discussion on Slack channel, the utility functions are being put in separate files. In this PR, these are `distance` functions in `distance.cpp` and `closestPointOnLine` functions in `closest_point.cpp`. I also put these under a `util` directory within `new_geom` to better organize it.
2. I made the new `Line` class constructor throw a `runtime_error` if it's called with two of the same `Point`s. Previously, this was handled by a `util` function `isDegenerate`, but I think an invalid `Line` should be disallowed in the first place.

### Testing Done

Passed all existing unit tests and added unit tests for new functions

### Resolved Issues

#1042

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by MathewMacDougall** [comment]
@garethellis0 so this shows up in your email.

We're going to do a first pass review (rather than wait for someone else to do the first review) since most people are busy during finals, and we don't want work to get held up unnecessarily.

File: src/software/new_geom/util/closest_point_test.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/util/closest_point_test.cpp]: Can these test names be more descriptive? Ideally they should state what the test is testing, eg. `point_far_from_line` or `point_close_to_line_but_not_on_line`

No need to worry about making the names too long, since they only need to be written once (it's not a variable that developers have to deal with) and makes reading and fixing tests much easier.

File: src/software/new_geom/util/distance.h
Comment (MathewMacDougall) [inline src/software/new_geom/util/distance.h:14]: @garethellis0 thoughts on if we want to duplicate javadoc comments for functions iike this where we simply flip the parameters? This is quite common with our utility function so I'm fine with them sharing the same comment and just being right next to one another.

eg.
```
/**
 * Finds the shortest distance between a Line and a Point
 *
 * @param first the Line
 * @param second the Point
 * @return the shortest distance between first and second
 */
double distance(const Line &first, const Point &second);
double distance(const Point &first, const Line &second);
```

File: src/software/new_geom/util/distance_test.cpp
Comment (MathewMacDougall) [inline src/software/new_geom/util/distance_test.cpp]: Same comment as above re: descriptive test names

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Looking good :+1:

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp
Comment (jsn0716) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp:10]: The interface for the new `closestPointOnLine` is the same as the old one, so it's not showing up in the diff but it is used.

File: src/software/ai/hl/stp/tactic/receiver_tactic.cpp
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/receiver_tactic.cpp:10]: Same as previous point, same interface so it's not showing in the diff

File: src/software/backend/input/network/filter/ball_filter.cpp
Comment (jsn0716) [reply src/software/backend/input/network/filter/ball_filter.cpp:324]: `toNormalUnitVector` produces a vector that points in a direction perpendicular to the regression line, so wouldn't `perpendicular` be necessary to make it point in the direction of regression line?

File: src/software/new_geom/util/closest_point.h
Comment (jsn0716) [reply src/software/new_geom/util/closest_point.h]: Good point, removed here: 3c5a97431b252

File: src/software/new_geom/util/closest_point.h
Comment (jsn0716) [reply src/software/new_geom/util/closest_point.h:13]: Added

File: src/software/new_geom/util/closest_point_test.cpp
Comment (jsn0716) [reply src/software/new_geom/util/closest_point_test.cpp]: Changed to be more descriptive and added a few more test cases

File: src/software/new_geom/util/distance_test.cpp
Comment (jsn0716) [reply src/software/new_geom/util/distance_test.cpp]: Changed to be more descriptive and added a few more test cases

File: src/software/new_geom/BUILD
Comment (garethellis0) [inline src/software/new_geom/BUILD]: Do we need all of `new_geom` to test this? We should make sure libraries are as small as possible so it's clear what's being tested. If we don't need all of `new_geom` here we should make a `closest_point` library and just test that (and have `new_geom` depend on it).

File: src/software/new_geom/line.h
Comment (garethellis0) [inline src/software/new_geom/line.h]: Returning an arbitrary 3-long array is a bit ambiguous, but I can see why we're using this here. Could we make it `protected` instead of `public`, then use a test-class that's a `friend` of this class? (I'm not being super detailed here, so feel free to ping me and I can explain exactly what to do in more detail).

This way we can use it for tests, but most users won't have access to it.

File: src/software/new_geom/BUILD
Comment (garethellis0) [reply src/software/new_geom/BUILD]: Same goes for `distance_test`, I'm really trying to avoid having a massive monolithic geom library like we currently do.

File: src/software/new_geom/util/distance.h
Comment (garethellis0) [reply src/software/new_geom/util/distance.h:14]: Agreed, this seems like a reasonable idea.

Comment (garethellis0) [reply src/software/new_geom/util/distance.h:14]: (as long as formatting doesn't insert an empty line between them).

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/software/backend/input/network/filter/ball_filter.cpp
Comment (MathewMacDougall) [reply src/software/backend/input/network/filter/ball_filter.cpp:324]: Ah you're right in that case. I read is as it converts the line to a normalized unit vector, where "normalized" means "unit vector" as well (because we use `norm` to create unit vectors in the `Vector` class).

This should be fine then.

File: src/software/new_geom/line.h
Comment (jsn0716) [reply src/software/new_geom/line.h]: This was made public so that utility functions like `distance` can access the coefficients. I could wrap those functions in a class and make them static instead, and make it a `friend` class along with the test class. @MathewMacDougall thoughts on this?

File: src/software/new_geom/BUILD
Comment (jsn0716) [reply src/software/new_geom/BUILD]: Added `closest_point` and `distance` as separate libraries, and `new_geom`, `closest_point_test`, and `distance_test` depend on them.

File: src/software/new_geom/util/distance.h
Comment (jsn0716) [reply src/software/new_geom/util/distance.h:14]: Done, clang_format doesn't seem to have any problems with it

File: src/software/new_geom/BUILD
Comment (garethellis0) [inline src/software/new_geom/BUILD]: If there are multiple libraries that depend on `line`, `point`, etc., then we should create libraries for `line`, `point`, etc. and then use `deps` to depend on them. This way we avoid unnecesarry re-compiling of files and make it way easier to manage changing dependencies in the future (consider the case where `line` depends on some new library).

File: src/software/new_geom/line.h
Comment (garethellis0) [reply src/software/new_geom/line.h]: We could also make `distance` a [friend function](https://www.tutorialspoint.com/cplusplus/cpp_friend_functions.htm) but not sure if that's a road we want to go down. @MathewMacDougall - your thoughts?

File: src/software/new_geom/geom_constants.h
Comment (garethellis0) [inline src/software/new_geom/geom_constants.h:1]: :scream:

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/software/new_geom/geom_constants.h
Comment (jsn0716) [reply src/software/new_geom/geom_constants.h:1]: Seems like this is not causing issues for my PR anymore (maybe after changes to BUILD), but isn't it safer to put `pragma once` anyways?

File: src/software/new_geom/BUILD
Comment (jsn0716) [reply src/software/new_geom/BUILD]: Updated BUILD!

File: src/software/new_geom/geom_constants.h
Comment (garethellis0) [reply src/software/new_geom/geom_constants.h:1]: Yup yup, this is absolutely correct. No header guard can result in nasty errors. Was just reacting at the fact that we didn't have this before :sweat_smile: 

See here for more info: https://www.learncpp.com/cpp-tutorial/header-guards/

File: src/software/new_geom/BUILD
Comment (garethellis0) [inline src/software/new_geom/BUILD]: In general, we try put libraries at the same level as the files they're composed of. Could we create a `BUILD` file in `util` and put this there (same goes for other libraries in `util` here).

File: src/software/new_geom/BUILD
Comment (garethellis0) [reply src/software/new_geom/BUILD]: Looking good, left one more comment below.

File: src/software/new_geom/line.h
Comment (MathewMacDougall) [reply src/software/new_geom/line.h]: hmmm yea I'd prefer we don't start making a bunch of `friend functions` since that could get complicated quickly (we might start wanting to do that a lot).

A wrapper `friend` class that makes the functions static also seems complicated.

@garethellis0 what if we returned a "less arbitrary" datatype instead? Maybe a struct with named coefficients or something? Is the main thing you don't like that the numbers in the array have no obvious "meaning", or that it's easy to go out of bounds?

File: src/software/new_geom/BUILD
Comment (jsn0716) [reply src/software/new_geom/BUILD]: The libraries and tests in `util` are now in the new `BUILD` file in `util`. I also made other geometry classes (`Ray`, `Polynomial`, etc.) into their own libraries except shapes since more are being added right now.

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada: - pending Mat's comments

**Comment by MathewMacDougall** [comment]
@garethellis0 waiting on your reply to one of my comments

File: src/software/new_geom/line.h
Comment (garethellis0) [reply src/software/new_geom/line.h]: I could live with returning a struct here, that was definitely one of the other options I was thinking of.

Comment (jsn0716) [reply src/software/new_geom/line.h]: Changed to return a struct

**Review by MathewMacDougall — DISMISSED** [review]

File: src/software/new_geom/line.cpp
Comment (garethellis0) [inline src/software/new_geom/line.cpp]: Could we change this to:
```
Coeffs returnme;
returnme.a = coeffs[0];
returneme.b = coeffs[1];
returneme.c =coeffs[2];
```
(also with a better name for `returnme`). If we were ever to change the order of the struct members, what is here currently would still compile :grimacing: 

Or alternatively we just store the coeffs in a struct internally as well? Unless there's another reason why not I'd suggest going this route.

Comment (jsn0716) [reply src/software/new_geom/line.cpp]: Not really any reason not to, changed the internal representation to struct

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by scveloso — APPROVED** [review]

**Pull request merged**
